Python rfactor

alexmiame opened the pull request:
This PR adds a native Python R factor calculation to viperleed. An effort is made to keep the code agnostic to the numerical library, such that the same routines can be used with NumPy and in JAX (used in the viperleed-jax plugin, see also #333).

alexmiame committed:
test: test rfactor file

alexmiame committed:
refactor: use src layout

alexmiame committed:
feat: rename to viperleed_jax

alexmiame committed:
fix: leftovers from rename merge

alexmiame committed:
fix: forgot to rename stuff

alexmiame committed:
Merge branch 'main' into improved_viperleed_compatibility

alexmiame committed:
fix: MS R-factor

alexmiame committed:
refactor: move interpolation code

alexmiame committed:
test: rfactor

alexmiame committed:
chore: module docstrings, authors and dates

Where not already noted in the file, creation dates were extracted from git via:
git log --follow --format=%ad --date default <FILE_PATH> | tail -1

alexmiame committed:
Merge pull request #2 from viperleed/improved_viperleed_compatibility

running from viperleed state and refactor

alexmiame committed:
Merge branch 'main' of https://github.com/viperleed/on-the-fly-deltas

alexmiame committed:
chore: reformat codebase

alexmiame committed:
style, doc: changes according to ruff linter

alexmiame committed:
refactor: move rfactor synonyms to rfactor module

alexmiame committed:
style,minor: redundant brackets

alexmiame committed:
style,doc: ruff formating

alexmiame committed:
chore,refactor: by ruff

alexmiame committed:
doc: by ruff

alexmiame committed:
style: ruff formatting

alexmiame committed:
test,chore: ruff formatting

alexmiame committed:
chore: remove unused code

alexmiame committed:
feat: per beam R-factors

alexmiame committed:
fix: unclosed bracet

alexmiame committed:
fix: missing argument in MS Rfactor

alexmiame committed:
chore: move constant to start of file

alexmiame committed:
fix: move it back

alexmiame committed:
feat: per_beam for R2, R1 and RZJ

alexmiame committed:
fix: NaNs in gradient of ZJ R factor

alexmiame committed:
feat: add Michaels new smooth R factor

alexmiame committed:
feat: kwargs for alpha/beta

alexmiame committed:
chore: remove RMS synonyms

alexmiame committed:
fix: shift interpolated away from minimum

fkraushofer committed:
fix: R_FACTOR_TYPE needs SpecialParameter default

The previous string default would cause R-factor calculations to fail because of int('pendry')

alexmiame committed:
Merge branch 'from_viperleed_jax' into python_rfactor

alexmiame committed:
feat: use numpy and scipy

alexmiame committed:
doc: adapt header

alexmiame committed:
feat: change jax and interpax to numpy/scipy

alexmiame committed:
feat: rfactor change jax for numpy

alexmiame committed:
doc: update header

alexmiame committed:
fix: remove non-scipy argument check

alexmiame committed:
feat,test: adapt r factor test

alexmiame committed:
feat: rfactor grouping function

alexmiame committed:
test: tests for grouped R factors

alexmiame committed:
feat,chore: change API to use groups

alexmiame committed:
chore: unused pendry R version

alexmiame committed:
feat,WIP: split rfactor into submodules

alexmiame committed:
feat: use dynamic numeric library

alexmiame committed:
chore: remove outdated integer V0r shift

alexmiame committed:
feat,chore: remove refactored rfactors

alexmiame committed:
fix: unused import

alexmiame committed:
chore,refactor: consistent R factor names

alexmiame committed:
feat: import R factors to init

alexmiame committed:
todo: remove TODO that was decided not to be pursued

alexmiame committed:
feat,minor: add more rfactor synonyms

alexmiame committed:
fix: file and function names

alexmiame committed:
test,fix: pendry_R name

alexmiame committed:
test,refactor: move grouping tests

alexmiame committed:
test,refactor: move rfactor tests

alexmiame committed:
doc: docstring

alexmiame committed:
fix: implement stop_gradient function

See also 068ea92af96ae83a37ace921308e0b6db115bfce

alexmiame committed:
fix: force numpy array

alexmiame committed:
chore,fix: kwargs for groups

alexmiame committed:
fix: calculation via bincount

fkraushofer commented on src/viperleed/calc/lib/rfactor/groups.py:
```diff
@@ -0,0 +1,59 @@
+    numerators : dnl.xp.ndarray
+        Array of R-factor numerators of shape (n_beams,)
+    denominators : dnl.xp.ndarray
+        Array of R-factor denominators of shape (n_beams,)
```
I don't understand this docstring, or really what the function does. What is a "R-factor numerator / denominator"? I guess it has something to do with how much data is in each beam, since it shows up in the weights somehow?

alexmiame replied:
For one beam, the various R factors are generally calculated as $`\int f(I_1, I_2) /  \int g(I_1, I_2)`$, where each integral runs over that beams energy range. For example, with Pendry's R factor, this would be $`f=(Y_1 -Y_2)^2`$ and $`g=(Y_1^2 + Y_2^2)`$.

To get the R factor over multiple beams $`g`$, you sum the numerators and denominators separately, and calculate the overall fraction $` \sum_g \int f(I_1, I_2) / \sum_g \int g(I_1, I_2)`$.
The group function takes the numerators and denominators for all beams and sums them to arbitrary subsets, including the special case of per-beam R factors and the overall one.

I can include this explanation as such in the docstring?

fkraushofer replied:
Oh, ok. Yeah, that makes sense. Yes, I think including a short version of this somewhere in the docstring would be good

fkraushofer replied:
Effectively, this ends up being a weighted average of individual R-factors, with the "weight" being the respective ranges of the integrals

alexmiame replied:
Yeah, pretty much.

fkraushofer commented on src/viperleed/calc/lib/rfactor/__init__.py:
```diff
@@ -0,0 +1,27 @@
+    R_pendry: ('pendry', 'r_p', 'rp', 'pendry r-factor', 'p'),
+    R_1: ('r1', 'r_1', 'r1 factor'),
+    R_2: ('r2', 'r_2', 'r2 factor'),
+    R_s: ('s', 'rs', 'r_s', 'smooth', 'schmid'),
```
I can't find at a glance where this is used. Is the check for these case-sensitive? If yes, then R_pendry should include r_pendry. If no (which would be better), then r_1, r_2, and r_s are redundant here

alexmiame replied:
So far, it's only used in viperleed-jax. However, since it's more general I thought it best to include it here.

alexmiame committed:
doc: docstring for R factor grouping

fkraushofer commented:
Only one minor comment, besides the docstring - looks good!

Presumably, this could now completely replace `viperleed.calc.sections.rfactor.run_new_rfactor`? How much hassle would it be to update that function to make use of these new ones, to avoid future confusion due to two different implementations? Since we already have the `R_FACTOR_LEGACY` parameter implemented, one could then immediately and easily use these.

alexmiame commented:
> Only one minor comment, besides the docstring - looks good!
> 
> Presumably, this could now completely replace `viperleed.calc.sections.rfactor.run_new_rfactor`? How much hassle would it be to update that function to make use of these new ones, to avoid future confusion due to two different implementations? Since we already have the `R_FACTOR_LEGACY` parameter implemented, one could then immediately and easily use these.

Hm, yes, I think the Fortran/F2Py R-factor version can be scraped. I can have a look at implementing it there.

alexmiame committed:
feat: add more dynamic functions

alexmiame committed:
feat: rfactor selection function

alexmiame committed:
feat: average_beam_array

moved from viperleed-jax

alexmiame committed:
feat: determine_integer_or_fractional

alexmiame committed:
feat(!): new R-factor calculation section

alexmiame committed:
fix: merge conflict and imports

alexmiame commented:
I've now added a first version of a standalone R-factor segment using the new functionality. Could you give it a try when you get the chance?
You can activate it by setting `R_FACTOR_LEGACY = False`.

fkraushofer commented:
Seems there is some issue:

```
STARTING SECTION: R-FACTOR CALCULATION
# WARNING: Using new R-factor calculation. This is still experimental!
Calculating R-factor type smooth.
dbg: Beam correspondence: (2, 1, 3, 3, 7, 6, 7, 6, 12, 11, 15, 15, 17, 17, 16, 16, 19, 18, 19, 18, 26, 25, 26, 25, 30, 30, 29, 29, 31, 31, 35, 34, 35, 34, 39, 39, 38, 38, 42, 41, 42, 41, 46, 45, 52, 52, 51, 51, 55, 55, 59, 58, 59, 58, 61, 60, 61, 60, 63, 63, 62, 62, 65, 64, 65, 64, 72, 72, 71, 71, 73, 73, 77, 76, 77, 76, 79, 79, 80, 80, 82, 81, 82, 81, 86, 85, 86, 85, 87, 87, 0, 0, 5, 5, 4, 4, 8, 8, 10, 9, 10, 9, 14, 14, 13, 13, 21, 21, 20, 20, 22, 22, 24, 23, 24, 23, 28, 28, 27, 27, 33, 32, 33, 32, 37, 37, 36, 36, 40, 40, 44, 43, 44, 43, 48, 48, 47, 47, 50, 50, 49, 49, 54, 53, 54, 53, 57, 57, 56, 56, 66, 66, 68, 68, 67, 67, 69, 69, 70, 70, 75, 74, 75, 74, 78, 78, 83, 83, 84, 84)
Best inner potential shift: -1.00 eV with R = 0.3338
### ERROR ### in run_sections:run_section:237
# Error in section R-FACTOR CALCULATION
#############
### ERROR ### in run_sections:section_loop:412
# Exception during viperleed.calc execution: 
#############
Traceback (most recent call last):
  File "/home/fs71704/fkraushofer/viperleed/viperleed/src/viperleed/calc/sections/run_sections.py", line 296, in section_loop
    run_section(sec, sl, rp)
  File "/home/fs71704/fkraushofer/viperleed/viperleed/src/viperleed/calc/sections/run_sections.py", line 218, in run_section
    rfactor.rfactor(sl, rp, index)
  File "/home/fs71704/fkraushofer/viperleed/viperleed/src/viperleed/calc/sections/rfactor.py", line 71, in rfactor
    return run_new_rfactor(sl, rp, for_error, name, theobeams, expbeams)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/fs71704/fkraushofer/viperleed/viperleed/src/viperleed/calc/sections/rfactor.py", line 229, in run_new_rfactor
    r_fac_integer, r_fac_fractional = r_func(
                                      ^^^^^^^
  File "/home/fs71704/fkraushofer/viperleed/viperleed/src/viperleed/calc/lib/rfactor/smooth.py", line 103, in R_s
    return R_pendry_from_y(y_exp, y_theo, energy_step, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/fs71704/fkraushofer/viperleed/viperleed/src/viperleed/calc/lib/rfactor/pendry.py", line 54, in R_pendry_from_y
    return group_rfactors(numerators, denominators, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/fs71704/fkraushofer/viperleed/viperleed/src/viperleed/calc/lib/rfactor/groups.py", line 74, in group_rfactors
    grouped_numerators = dnl.bincount(
                         ^^^^^^^^^^^^^
  File "/home/fs71704/fkraushofer/.conda/envs/on-the-fly-devel/lib/python3.12/site-packages/jax/_src/numpy/lax_numpy.py", line 3012, in bincount
    raise ValueError("shape of weights must match shape of x.")
ValueError: shape of weights must match shape of x.
```

edit: PARAMETERS:
```
R_FACTOR_TYPE = smooth
R_FACTOR_LEGACY = False
```

edit2: maybe also relevant: some beams were discarded:
```
# WARNING: The following beams contain no data in the given energy range and will be ignored: ( 2| 5), ( 4|-4), ( 3|-5), (-3| 5), ( 0|-6), ( 0| 6), ( 5| 3), (9/2| 4 ), ( 6| 0)
```

fkraushofer commented:
The same error also occurs with viperleed-jax backend when using the current newest versions of both viperleed and viperleed-jax:

```
Optimizer SLSQP finished with best R = 0.2927
dbg: Optimizer result:
OptimizationHistory (SLSQP) R:   0.2927

dbg: Wrote to POSCAR_TLOpt_0_SLSQP_intermediate successfully
### ERROR ### in run_sections:run_section:237
# Error in section SEARCH
#############
### ERROR ### in run_sections:section_loop:412
# Exception during viperleed.calc execution: 
#############
Traceback (most recent call last):
  File "/home/fs71704/fkraushofer/viperleed/viperleed/src/viperleed/calc/sections/run_sections.py", line 296, in section_loop
    run_section(sec, sl, rp)
  File "/home/fs71704/fkraushofer/viperleed/viperleed/src/viperleed/calc/sections/run_sections.py", line 227, in run_section
    vlj_search(sl, rp)
  File "/home/fs71704/fkraushofer/viperleed/viperleed/src/viperleed/calc/vlj/search.py", line 182, in vlj_search
    integer_fractional_R = calculator.R(
                           ^^^^^^^^^^^^^
  File "/home/fs71704/fkraushofer/viperleed/viperleed-jax/src/viperleed_jax/tensor_calculator.py", line 649, in R
    return calc_r_factor(
           ^^^^^^^^^^^^^^
  File "/home/fs71704/fkraushofer/viperleed/viperleed-jax/src/viperleed_jax/tensor_calculator.py", line 948, in calc_r_factor
    return rfactor_func(
           ^^^^^^^^^^^^^
  File "/home/fs71704/fkraushofer/viperleed/viperleed/src/viperleed/calc/lib/rfactor/smooth.py", line 103, in R_s
    return R_pendry_from_y(y_exp, y_theo, energy_step, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/fs71704/fkraushofer/viperleed/viperleed/src/viperleed/calc/lib/rfactor/pendry.py", line 54, in R_pendry_from_y
    return group_rfactors(numerators, denominators, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/fs71704/fkraushofer/viperleed/viperleed/src/viperleed/calc/lib/rfactor/groups.py", line 74, in group_rfactors
    grouped_numerators = dnl.bincount(
                         ^^^^^^^^^^^^^
  File "/home/fs71704/fkraushofer/.conda/envs/on-the-fly-devel/lib/python3.12/site-packages/jax/_src/numpy/lax_numpy.py", line 3012, in bincount
    raise ValueError("shape of weights must match shape of x.")
ValueError: shape of weights must match shape of x.
--------------------
For simplicity, JAX has removed its internal frames from the traceback of the following exception. Set JAX_TRACEBACK_FILTERING=off to include these.
```

with PARAMETERS:
```
BACKEND search = VIPERLEED-JAX
R_FACTOR_TYPE = smooth
VLJ_CONFIG = t-leed-l_max 12
VLJ_BATCH = energies 59, atoms 74
VLJ_ALGO = SLSQP
```

No beams dropped in this case, so I think we can rule out the "edit2" above as the source of the problem

alexmiame commented:
I'm investigating this atm. Part of the problem seems to be again when I call getBeamCorrespondence, which alters the beamlist.
Maybe we should consider calling that already at initialization time, so the indexing cannot change during the run.

I'll get back to you, once I have a fix.

alexmiame committed:
fix: spline coefficients shape

alexmiame committed:
fix: call getBeamCorrespondence earlier

Since alters the beamlists in-place

alexmiame committed:
fix: correct integer-or-fractional mask size

Required size is length of expbeams, not theobeams

alexmiame commented:
Alright, I think this should do it; at least it works on my side now. Let me know.

fkraushofer commented on src/viperleed/calc/lib/rfactor/__init__.py:
```diff
@@ -0,0 +1,36 @@
+from .zannazi_jona import R_zj
+
+_R_FACTOR_SYNONYMS = {
+    R_pendry: ('pendry', 'r_p', 'rp', 'pendry r-factor', 'p'),
```
```suggestion
    R_pendry: ('pendry', 'r_p', 'r_pendry', 'rp', 'pendry r-factor', 'p'),
```

fkraushofer commented on src/viperleed/calc/lib/rfactor/__init__.py:
```diff
@@ -0,0 +1,36 @@
+    R_pendry: ('pendry', 'r_p', 'rp', 'pendry r-factor', 'p'),
+    R_1: ('r1', 'r_1', 'r1 factor'),
+    R_2: ('r2', 'r_2', 'r2 factor'),
+    R_s: ('s', 'rs', 'r_s', 'smooth', 'schmid'),
```
```suggestion
    R_s: ('s', 'rs', 'r_s', 'r_smooth', 'smooth', 'schmid'),
```

alexmiame committed:
feat: add pendry synonym

Co-authored-by: Florian Kraushofer <66359721+fkraushofer@users.noreply.github.com>

alexmiame committed:
feat: add r_smooth synonym

Co-authored-by: Florian Kraushofer <66359721+fkraushofer@users.noreply.github.com>

alexmiame committed:
fix: apply stop_gradient

fkraushofer committed:
feat: Implement basic plotting for new r-factor

R-factor analysis file is still missing, but this restores basic plotting functionality for the new R-factor function, including when using R_s.

fkraushofer committed:
fix: type conversion

fkraushofer committed:
minor: more detail in debug messages

fkraushofer committed:
feat: write_Rfactorpdf_from_splines

Functionality was getting bigger, moved to a new function in iorfactor

fkraushofer committed:
feat: spline cashing, built-in theory shifts

Two optimizations for the python-based R-factor:
1. When shifting V0r, do not re-build splines at the shifted coordiantes; instead, use the existing splines and evaluate them at the shifted coordinates (the splines cover the whole energy range in any case).
2. CashedSpline: Avoid re-evaluating the derivatives and values repeatedly on the same spline (e.g. also when scanning V0r, expecially the experimental data is always the same). Instead, keep the previous evaluation in memory.

fkraushofer committed:
fix: explicit check in CashedSpline

checking id(x) is not enough; actually compare the object

fkraushofer committed:
minor: add theo_shift in R_ms

missed in previous commit

fkraushofer committed:
minor: debug message

give some additional progress info during rfactor calculation

fkraushofer committed:
feat: accept pre-computed arrays

- change 'theo' and 'exp' nomenclature in R-factor functions to 'data_1' and 'data_2'
- add doc strings
- drop R_ms (outdated version of R_s)
- allow passing *either* splines *or* pre-computed data arrays (already on the right grid) to the R-factor functions. The latter is more JAX-friendly for the experimental data, since we would otherwise have to deal with ragged arrays, where spline interpolation cannot be cleanly vectorized.

Note that this changes the call signatures of the R-factor functions, and requires also updating the JAX plugin.

fkraushofer committed:
fix: typo from last commit

fkraushofer committed:
fix: another issue from previous commits

fkraushofer committed:
feat: Update to access R_S from TensErLEED

TODO: update in base branch
TODO: update docs
TODO: Check TensErLEED version, default to R_P if version too old
Co-Authored-By: schmid-iap <82087012+schmid-iap@users.noreply.github.com>

fkraushofer committed:
fix: Check TL_VERSION if R_smooth is available

fkraushofer committed:
Merge branch '0.14.2' into python_rfactor

fkraushofer committed:
merge: Manual cherry-pick of parts of #526

Gets mainly r_factor_type as special parameter, and some minor changes, into the python_rfactor branch, to be merged into 14.2 independent of viperleed-jax-plugin.

fkraushofer committed:
Merge: More manual stuff from #526

Forgotten in last commit. This should be all.

fkraushofer committed:
Update _checksums.dat for TensErLEED v2.1.0

Add checksums for new TensErLEED version v2.1.0

fkraushofer committed:
fix: update rfactor tests

fkraushofer committed:
feat: add test data for rfactor regression tests

fkraushofer committed:
fix: arguments for np.clip

should also be fine for jax.numpy.clip

fkraushofer committed:
fix: possible divide-by-zero in groups

fkraushofer committed:
test,refactor: Tests for python rfactor

fkraushofer committed:
test: R_FACTOR_TYPE special parameter

fkraushofer committed:
chore: delete obsolete writeRfactorPdf_new

Old and not very functional implementation, never used. Replaced by write_Rfactorpdf_from_splines

fkraushofer committed:
fix: name

fkraushofer committed:
fix: name,authors

fkraushofer commented:
@michele-riva please have a quick look when you have a minute. In principle there was 4 eyes on this already, but since both @alexmiame and me worked on it your review is required.

Tests pass and have good coverage. I also added a regression test with real data from hematite (1x1) for all R-factor types, and did some "real-world" full executions of `viperleed calc` (`RUN = 1-3 1`, and separate `RUN = 11` both with legacy R-factors and without). All looks good.

There are some changes in the files which already encroach a bit on the jax-plugin branch / #333, since this was first pushed there. I did my best to disentangle things, but stuff like the dynamic library is already in here, since it would be unreasonably complicated to try and disentangle this. However, this branch runs fine without jax support.

This also conflates a bit the python R-factor functionality by @alexmiame and the new Fortran implementation of R_smooth by @schmid-iap - that's my bad, sorry. In any case, this branch already supports the R_smooth implementation in TensErLEED v2.1.0. @schmid-iap requested some further minor changes on the `viperleed calc` side, but these I will do as a separate branch.

fkraushofer committed:
feat: label R-factor correctly in errorcalc output

Distinguish whether R_Pendry or R_Smooth was used

fkraushofer committed:
refactor: rename whichR to which_beams

whichR is generally used for the type of R-factor used (Pendry/Smooth/etc), but here it distinguishes which group of beams to look at (all, integer, or fractional). New variable name which_beams should be more clear.

fkraushofer committed:
Merge pull request #539 from viperleed/r_factor_cleanup

R-factor cleanup

fkraushofer committed:
feat: use new python rfactor in from_ase

fkraushofer committed:
chore: delete viperleed.extensions.rfactor

@alexmiame should we keep the other extensions, or were they only used here?

fkraushofer committed:
fix: beam matching without explicit correspondence

fkraushofer committed:
fix: IV plots in from_ase

fkraushofer committed:
test: basic tests for from_ase rfactor

michele-riva commented on src/viperleed/calc/classes/rparams/special/r_factor_type.py:
```diff
@@ -0,0 +1,197 @@
+"""Module l_max of viperleed.calc.classes.rparams.special.
+
+Defines the LMax class, a convenience container for parameter LMAX.
+"""
```
Module docstring does not match contents

michele-riva commented on src/viperleed/calc/classes/rparams/special/r_factor_type.py:
```diff
@@ -0,0 +1,197 @@
+    def __init__(self, value):
+        r_fac_id, name = self._coerce(value)
+        self.id = r_fac_id
+        self.name = name
```
Not sure we ever use these attributes outside this class, but leaving them editable does not seem like a good idea.

michele-riva commented on src/viperleed/calc/classes/rparams/special/r_factor_type.py:
```diff
@@ -0,0 +1,197 @@
+    """Exception for RFactorType-related errors."""
+
+
+class RFactorType(SpecialParameter, param='R_FACTOR_TYPE'):
```
Is there a reason not to use a real Enum for this? It would allow using "is" comparisons rather than having to resort on converting to string and comparing each time.

michele-riva commented on src/viperleed/calc/files/parameters/interpret.py:
```diff
@@ -182,6 +186,9 @@ def _wrapper_method(self, assignment, **kwargs):
     return cls
 
 
+_BELOW_DEBUG = 2
+
+
```
I think I had unified these levels in 4db279d. This seems to be reverting part of those changes.

michele-riva commented on src/viperleed/calc/files/parameters/interpret.py:
```diff
@@ -278,6 +285,13 @@ def _update_param_order(self):
             if p in self.rpars.readParams and p not in self._param_names
             )
 
+    def _parse_simple_bool(self, val):
+        val_lower = val.lower()
+        for key, values in self.bool_synonyms.items():
+            if val_lower in values:
+                return key
+        raise ValueError(f"Unrecognized boolean value: {val!r}")
+
```
This helper seems unused

michele-riva commented on src/viperleed/calc/files/ioerrorcalc.py:
```diff
@@ -319,6 +318,15 @@ def make_errors_figs(errors, formatting=None):
+        varlabel = '$R_S + var(R_S)$'
+    else:
+        ylabel = 'R-factor'   # this should not happen, just a contingency
+        varlabel = '$R + var(R)$'
```
maybe log some warning? Or one could even be stricter and raise a ValueError as this seems like a misuse of the argument that may go unnoticed

michele-riva commented on src/viperleed/calc/files/iorfactor.py:
```diff
@@ -468,12 +473,31 @@ def writeWEXPEL(sl, rp, theobeams, filename="WEXPEL", for_error=False):
+        output += ' WR=      1.,0.,0.,\n'
+    elif str(rp.R_FACTOR_TYPE) == 'zj':
+        output += ' WR=      0.,1.,0.,\n'
+    elif str(rp.R_FACTOR_TYPE) == 'smooth':
+        if rp.TL_VERSION < Version('2.1.0'):  # No R_S implemented
+            output += ' WR=      0.,0.,1.,\n'
+            logger.warning(
+                'Smooth R-factors are not supported outside of the search for '
+                'TensErLEED versions <2.1. Falling back on Pendry R-factor. '
+                'Resulting R-factors may differ from those calculated during '
+                'the search.'
+                )
+        else:
+            # rfactor.f changes 4th weight to 1
+            output += ' WR=      0.,0.,0.,\n'
```
Perhaps nicer to flatten this nested if out?

michele-riva commented on src/viperleed/calc/files/iorfactor.py:
```diff
@@ -468,12 +473,31 @@ def writeWEXPEL(sl, rp, theobeams, filename="WEXPEL", for_error=False):
-        output += ' WR=      0.,0.,1.,\n'  # Pendry
-    elif rp.R_FACTOR_TYPE == 2:
-        output += ' WR=      1.,0.,0.,\n'  # R2
+    if str(rp.R_FACTOR_TYPE) == 'pendry':
```
Are all these str conversions needed? `__eq__` does attempt conversion from string

michele-riva replied:
Applies to more modules below

alexmiame replied:
TODO: behavior (may) depend(s) on wether we make RfactorType into an Enum

michele-riva commented on src/viperleed/calc/files/iorfactor.py:
```diff
@@ -966,6 +910,87 @@ def beamlist_to_array(beams):
+
+    Parameters
+    ----------
+    rp : Rparams
```
```suggestion
    rpars : Rparams
```

fkraushofer committed:
feat: Make python rfactor the default

fkraushofer committed:
docs: update R-factor parameter docs

fkraushofer committed:
chore,doc: remove abstracts

not used, just takes up space and prone to errors

fkraushofer committed:
doc: Describe R_S

Co-Authored-By: schmid-iap <82087012+schmid-iap@users.noreply.github.com>

fkraushofer committed:
doc: Resolve todo

michele-riva commented on doc/content/calc/parameters/rfactorlegacy.rst:
```diff
@@ -5,12 +5,18 @@
 R_FACTOR_LEGACY
 ===============
 
+.. versionchanged:: 0.15.0
+   Since version 0.15.0, the default is to use the new python-native |R factor|
+   calculation.
+
+
```
I think we should perhaps maintain the default to `True` until we have merged the jax support. This is so that we limit discrepancy of values between search results and r-factor ones, since we de-facto now only support the TensErLEED backend.

In the jax-support branch we can adapt the default depending on which search backend is selected.

fkraushofer replied:
Fine to not change it until we merge viperleed-jax. I wouldn't want to have it adapt dynamically to the search backend though: Much better to have slight discrepancies between refcalc- and search-rfactors imo than to have them between different refcalc in the same history (which would happen if users specify search backend only once they get to that part, or if they comment it out at some point because they're not doing a search - has happened to me). So, for consistency's sake I would argue we should have it either as default, or not.

michele-riva commented on doc/content/calc/sections/r-factor_calculation.rst:
```diff
@@ -14,78 +14,90 @@ the |R factor| between the calculated and experimental |IV| curves.
+    Using |RP| is the **default setting**.
+    For structure optimization, we recommend using |RS|, since it is better suited
+    as a target function for minimization than |RP| (|RP| is noisy and
+    has many local minima in the vicinity to the global minimum).
```
```suggestion
    has many local minima in the vicinity of the global minimum).
```

michele-riva commented on doc/content/calc/sections/r-factor_calculation.rst:
```diff
@@ -14,78 +14,90 @@ the |R factor| between the calculated and experimental |IV| curves.
-    Y(E) = \frac{I(E)/I'(E)}{[I(E)/I'(E)]^2 + V_{0\mathrm{i}}^2}.
-
-The beam intensities enter Eq. :eq:`eq_Y_def` via their logarithmic
+The :math:`Y` functions in Eq. :eq:`eq_RP_def` are based on logarithmic
```
```suggestion
The :math:`Y` functions in Eq. :eq:`eq_RP_def` are based on the logarithmic
```

michele-riva commented on doc/content/calc/sections/r-factor_calculation.rst:
```diff
@@ -14,78 +14,90 @@ the |R factor| between the calculated and experimental |IV| curves.
 differences; artificially increasing the intensity at minima thus gives
 smaller |RP| values. Some LEED programs apply such a smoothing; in those
-cases smaller |R factor|\ s than those obtained with ViPErLEED will be
+cases smaller |R factors| than those obtained with ViPErLEED will be
```
```suggestion
cases, smaller |R factors| than those obtained with ViPErLEED will be
```

michele-riva commented on src/viperleed/calc/classes/rparams/rparams.py:
```diff
@@ -139,8 +139,8 @@ def __init__(self):
         self.RUN = self.get_default('RUN')        # what segments should be run
-        self.R_FACTOR_LEGACY = True # use old runtime-compiled R-factor calculation
-        self.R_FACTOR_TYPE = 1  # 1: Pendry, 2: R2, 3: Zanazzi-Jona
+        self.R_FACTOR_LEGACY = False  # use old runtime-compiled R-factor calculation
```
See my note in the docs

michele-riva commented on src/viperleed/calc/files/iorfactor.py:
```diff
@@ -966,6 +910,87 @@ def beamlist_to_array(beams):
+    theo_spline : scipy.interpolate.PPoly or jaxlib.xla_extension.ArrayImpl
+        interpolation splines of theoretical I(V) data
+    exp_spline : scipy.interpolate.PPoly or jaxlib.xla_extension.ArrayImpl
+        interpolation splines of experimental I(V) data
```
I guess technically these just need to be callables with a single argument that return I(V) curves when evaluated on a grid of energies

michele-riva commented on src/viperleed/calc/files/iorfactor.py:
```diff
@@ -966,6 +910,87 @@ def beamlist_to_array(beams):
+        the grid on which to evaluate the splines
+    r_fac_per_beam : np.array of float
+        the R-factors per beam, in the same order as the beam data
+    outname : kwarg, str
```
```suggestion
    outname : str, optional
```

michele-riva commented on src/viperleed/calc/files/iorfactor.py:
```diff
@@ -966,6 +910,87 @@ def beamlist_to_array(beams):
+        name of the file (with or without extension) to which the plots
+        will be saved.
+        default: 'Rfactor_plots.pdf'
+    analysis_file : kwarg, string
```
```suggestion
    analysis_file : str, optional
```

michele-riva commented on src/viperleed/calc/files/iorfactor.py:
```diff
@@ -966,6 +910,87 @@ def beamlist_to_array(beams):
+    '''
+    Creates a single PDF file containing the plots of R-factors, using plot_iv.
+    If analysis_file is defined, a second 'analysis' PDF will be generated.
+    # TODO: Actually generate analysis files! #
```
Dosctring 1st line in imperative mode, one empty line between that and more details. If more details are present (like here) 1st line should be max 72 chars.

michele-riva commented on src/viperleed/calc/files/iorfactor.py:
```diff
@@ -966,6 +910,87 @@ def beamlist_to_array(beams):
+        The object holding information about the current PARAMETERS.
+        Attributes used: PLOT_IV, expbeams
+    theo_spline : scipy.interpolate.PPoly or jaxlib.xla_extension.ArrayImpl
+        interpolation splines of theoretical I(V) data
```
```suggestion
        Interpolation splines of theoretical I(V) data
```

same for all the others below concerning capitazation

michele-riva commented on src/viperleed/calc/files/iorfactor.py:
```diff
@@ -966,6 +910,87 @@ def beamlist_to_array(beams):
+    -------
+    None
+
+    '''
```
```suggestion
    None.
    '''
```

michele-riva commented on src/viperleed/calc/files/iorfactor.py:
```diff
@@ -966,6 +910,87 @@ def beamlist_to_array(beams):
+        outname='Rfactor_plots.pdf',
+        analysis_file='',
+        ):
+    '''
```
Docstring with triple-double quotes. We reserve triple-single for multiline string literals.

michele-riva commented on src/viperleed/calc/files/iorfactor.py:
```diff
@@ -966,6 +910,87 @@ def beamlist_to_array(beams):
+        plot_intens_exp = intens_exp[m] / max(intens_exp[m])
+        xy_theo.append(np.column_stack([plot_grid, plot_intens_theo]))
+        xy_exp.append(np.column_stack([plot_grid, plot_intens_exp]))
+    rfac_str = ["R = {:.4f}".format(r) for r in r_fac_per_beam]
```
```suggestion
    rfac_str = [f'R = {r:.4f}' for r in r_fac_per_beam]
```

michele-riva commented on src/viperleed/calc/files/iorfactor.py:
```diff
@@ -966,6 +910,87 @@ def beamlist_to_array(beams):
+        xy_theo.append(np.column_stack([plot_grid, plot_intens_theo]))
+        xy_exp.append(np.column_stack([plot_grid, plot_intens_exp]))
+    rfac_str = ["R = {:.4f}".format(r) for r in r_fac_per_beam]
+    labelstyle = "overbar" if rpars.PLOT_IV["overbar"] else "minus"
```
```suggestion
    labelstyle = 'overbar' if rpars.PLOT_IV['overbar'] else 'minus'
```

michele-riva commented on src/viperleed/calc/files/iorfactor.py:
```diff
@@ -966,6 +910,87 @@ def beamlist_to_array(beams):
+            legends=['Theoretical', 'Experimental'],
+            labels=labels,
+            annotations=rfac_str,
+            formatting=rpars.PLOT_IV
```
```suggestion
            formatting=rpars.PLOT_IV,
```

michele-riva commented on src/viperleed/calc/files/iorfactor.py:
```diff
@@ -966,6 +910,87 @@ def beamlist_to_array(beams):
+            annotations=rfac_str,
+            formatting=rpars.PLOT_IV
+            )
+    logger.debug(
```
Perhaps this should be info?

michele-riva commented on src/viperleed/calc/files/iorfactor.py:
```diff
@@ -966,6 +910,87 @@ def beamlist_to_array(beams):
+        intens_theo = theo_data[:, i]
+        intens_exp = exp_data[:, i]
+        # keep only points where BOTH are finite (non-NaN, non-inf)
+        m = np.isfinite(intens_theo) & np.isfinite(intens_exp)
```
m -> mask?

or any other renaming that makes it clearer what it is (e.g., `finite`)

michele-riva commented on src/viperleed/calc/files/iosearch.py:
```diff
@@ -91,7 +91,7 @@ def readSDTL_blocks(content, whichR=0, print_info=False, n_expect=0):
     content : str
         A block of data as read from SD.TL.
-    whichR : int, optional
+    which_group : int, optional
```
```suggestion
    which_beams : int, optional
```

or rename all to `which_group`

michele-riva replied:
Also, would be good to list which values correspond to which and what is the default

michele-riva commented on tests/calc/lib/rfactor/conftest.py:
```diff
@@ -0,0 +1,197 @@
+_MOCK_COS_INTENSITY = np.cos(_MOCK_ROUGH_E_AXIS / (3.0 * np.pi)) ** 2
+_MOCK_CONST_INTENSITY = np.ones_like(_MOCK_ROUGH_E_AXIS)
+
+TESTS_DIR = Path(__file__).resolve().parents[3]
+BEAM_DATA_DIR = TESTS_DIR / "_test_data" / "beam-data"
```
better import the _test_data path from tests/helpers

michele-riva commented on tests/calc/lib/rfactor/conftest.py:
```diff
@@ -0,0 +1,197 @@
+
+
+def pytest_addoption(parser):
+    parser.addoption(
+        "--update-rfactor-expected",
+        action="store_true",
+        default=False,
+        help="Regenerate frozen expected values for real-data rfactor regression tests.",
+    )
```
Is this going to create a new .expected.json? Perhaps would be good to mention this in the help.

michele-riva commented on tests/calc/lib/rfactor/conftest.py:
```diff
@@ -0,0 +1,197 @@
+"""Shared fixtures and helpers for rfactor tests."""
```
Perhaps nice to to swap double quotes for single ones in string literals for consistency with other tests.

michele-riva commented on tests/calc/lib/rfactor/conftest.py:
```diff
@@ -0,0 +1,197 @@
+
+TESTS_DIR = Path(__file__).resolve().parents[3]
+BEAM_DATA_DIR = TESTS_DIR / "_test_data" / "beam-data"
+EXP_BEAMS_FILE = BEAM_DATA_DIR / "EXPBEAMS-fe2o3-012-1x1.csv"
+THEO_BEAMS_FILE = BEAM_DATA_DIR / "THEOBEAMS-fe2o3-012-1x1.csv"
+BEAMCORR_FILE = BEAM_DATA_DIR / "beamcorr-fe2o3-012-1x1.json"
+EXPECTED_FILE = BEAM_DATA_DIR / "rfactor-fe2o3-012-1x1.expected.json"
```
For easier extensibility, consider whether it would be nicer to have these files further structured into a top folder with the name of the system and contents names just "EXPBEAMS.csv", etc.

michele-riva commented on tests/calc/test_from_ase.py:
```diff
@@ -38,8 +39,18 @@
 
 
 ASE_DATA = TEST_DATA / 'from_ase'
+BEAM_DATA_DIR = TEST_DATA / "beam-data"
```
```suggestion
BEAM_DATA_DIR = TEST_DATA / 'beam-data'
```

Single quotes also below for consistency

michele-riva commented on src/viperleed/calc/lib/rfactor/groups.py:
```diff
@@ -0,0 +1,85 @@
+        Array of R-factor numerators of shape (n_beams,)
+    denominators : dnl.xp.ndarray
+        Array of R-factor denominators of shape (n_beams,)
+    groups : None | "beam" | array-like of int, optional
```
```suggestion
    groups : None or "beam" or array-like of int, optional
```

michele-riva commented on src/viperleed/calc/lib/rfactor/groups.py:
```diff
@@ -0,0 +1,85 @@
+    num_groups : int, optional
+        Number of groups, required if groups is array-like of int.
+        Must be specified to ensure compatibility with just-in-time
+        compilation and static array shapes.
```
The docstring seems to imply that this is necessary if an array is passed, but it is silently set in the function. Is this really needed then?

alexmiame replied:
Yes. It will work in Numpy and non-compiled JAX, but will fail if used without in JIT compiled JAX because the array shape is not static if it depends on another value. I will adapat the docstring.

michele-riva commented on src/viperleed/calc/lib/rfactor/groups.py:
```diff
@@ -0,0 +1,85 @@
+
+    mask = grouped_denominators != 0
+    safe_denominators = dnl.xp.where(mask, grouped_denominators, 1.0)
+    result = dnl.xp.where(mask, grouped_numerators / safe_denominators, dnl.xp.nan)
```
Probably the docstring should mention this masking in the Returns documentation.
NB:
normally (-1/0, 0/0, +1/0) -> (-inf, nan, +inf)
however, all three above go to nan in this implementation

alexmiame replied:
Technically true, although returning NaNs for invalid R factors makes more sense.
I'll add a note to the docstring, even if it's a very specific technicallity.

michele-riva commented on src/viperleed/calc/lib/rfactor/__init__.py:
```diff
@@ -0,0 +1,36 @@
+from .smooth import R_s
+from .zannazi_jona import R_zj
+
+_R_FACTOR_SYNONYMS = {
+    R_pendry: ('pendry', 'r_p', 'r_pendry', 'rp', 'pendry r-factor', 'p'),
+    R_1: ('r1', 'r_1', 'r1 factor'),
+    R_2: ('r2', 'r_2', 'r2 factor'),
+    R_s: ('s', 'rs', 'r_s', 'r_smooth', 'smooth', 'schmid'),
+    R_zj: (
+        'zj',
+        'zj factor',
+        'zannazi',
+        'zannazi jona',
+        'zannazi-jona',
+    ),
+}
```
Looks like we have a duplicate list of aliases for these R factors (one is in the special parameters). I would suggest to keep only one of these to avoid the maintenance burden.

michele-riva commented on src/viperleed/calc/lib/rfactor/__init__.py:
```diff
@@ -0,0 +1,36 @@
+from .r_1 import R_1
+from .r_2 import R_2
+from .smooth import R_s
+from .zannazi_jona import R_zj
```
I think we tried to remove all the imports in __init__.py files as they can very easily lead to circular import problems. If any of these modules also imports from the package this __init__ will be partly initialized. It does not cause problems right now, but it may in the future. If the select_rfactor function is useful anywhere, perhaps it should go to a dedicated module.

michele-riva commented on src/viperleed/calc/lib/rfactor/pendry.py:
```diff
@@ -0,0 +1,115 @@
+from .utils import nansum_trapezoid, shift_theo_intensity_non_negative
+
+
+def R_pendry(
```
I would suggest to use small "r" for all of these to better conform with python style.

alexmiame replied:
Not sure, I think this may be a case where it's better for clarity to use the upper case. What do you think @fkraushofer?

fkraushofer replied:
Yeah, I feel like it's one case where one could make a good argument why uppercase R is ok. But, I'm fine with having it lowercase as well (we also have the y-functions lowercase in general). No particular preference, as long as we keep it consistent.

alexmiame replied:
Hm, good point about the Y functions. I guess it's clear enough what it is since it's imported from the `rfactor` module. I'll change it to lower case.
(That also gets rid of the annoying linter warning)

michele-riva commented on src/viperleed/calc/lib/rfactor/pendry.py:
```diff
@@ -0,0 +1,115 @@
+    **kwargs,
+    ):
+    """
+    Prendry R-factor, see
```
```suggestion
    """Calculate Pendry's R factor between I(V) curves.

    See
```

michele-riva commented on src/viperleed/calc/lib/rfactor/pendry.py:
```diff
@@ -0,0 +1,115 @@
+    Uses two sets of beam data, either passed as splines or as already
+    calculated arrays of intensities and derivatives. When comparing
+    experimental and theoretical data, the theoretical data should be the
+    second dataset, as this is also shifted to correct for non-negative values.
```
```suggestion
    second dataset, as this is also shifted to correct for non-positive values.
```

I guess?

michele-riva commented on src/viperleed/calc/lib/rfactor/pendry.py:
```diff
@@ -0,0 +1,115 @@
+    Parameters
+    ----------
+    v0_imag : float
+        The imaginary part of the inner potential.
```
I guess the absolute value

alexmiame replied:
Doesn't matter, it's only used squared.

michele-riva commented on src/viperleed/calc/lib/rfactor/pendry.py:
```diff
@@ -0,0 +1,115 @@
+    data_spline_2=None,
+    data_and_derivatives_2=None,
+    shift_2nd_spline=0.0,   # only available if passed as spline
+    **kwargs,
```
Missing from parameters doc

michele-riva commented on src/viperleed/calc/lib/rfactor/pendry.py:
```diff
@@ -0,0 +1,115 @@
+        data_1_intensity = data_spline_1(energy_grid)
+        data_1_derivative = data_1_deriv_spline(energy_grid)
+    else:
+        data_1_intensity, data_1_derivative, _ = data_and_derivatives_1
```
This block is somewhat repeated for the second dataset. Perhaps worth factoring it out into a private? A similar logic is also used in all the other R-factors. Perhaps worth even extracting this into a utils

michele-riva commented on src/viperleed/calc/lib/rfactor/pendry.py:
```diff
@@ -0,0 +1,115 @@
+
+def y_pendry(intensity, intensity_derivative, v0_imag):
+    intens_deriv_ratio = intensity / intensity_derivative
+    return intens_deriv_ratio / (intens_deriv_ratio**2 + v0_imag**2)
```
Perhaps it is numerically a bit more stable to avoid the first division?

```suggestion
    numerator = intensity * intensity_derivative
    denominator = intensity**2 + v0_imag**2 * intensity_derivative**2
    return numerator / denominator
```

michele-riva commented on src/viperleed/calc/lib/rfactor/pendry.py:
```diff
@@ -0,0 +1,115 @@
+    return R_pendry_from_y(y_1, y_2, energy_step, **kwargs)
+
+
+def R_pendry_from_y(y_1, y_2, energy_step, **kwargs):
```
missing docstring. Same for y_pendry below.

michele-riva commented on src/viperleed/calc/lib/rfactor/pendry.py:
```diff
@@ -0,0 +1,115 @@
+"""viperleed.calc.lib.rfactor.pendry."""
```
```suggestion
"""Module pendry of viperleed.calc.lib.rfactor."""
```

michele-riva commented on src/viperleed/calc/lib/rfactor/r_1.py:
```diff
@@ -0,0 +1,88 @@
+"""viperleed.calc.lib.rfactor.r_1."""
```
See also my comments on the pendry module. Most of them apply here too.

michele-riva commented on src/viperleed/calc/lib/rfactor/r_1.py:
```diff
@@ -0,0 +1,88 @@
+    **kwargs,
+    ):
+    """
+    R2 R-factor. Uses two sets of beam data, either passed as splines or as
```
Docstring and errors all refer to this as R2 though it is R1

michele-riva commented on src/viperleed/calc/lib/rfactor/pendry.py:
```diff
@@ -0,0 +1,115 @@
+        Evaluate the 2nd spline on a shifted energy grid. This is meant for
+        testing V0r variations. Note that this is NOT available when the 2nd
+        dataset is passed as data_and_derivatives.
+    """
```
Missing: Raises doc

michele-riva commented on src/viperleed/calc/lib/rfactor/r_2.py:
```diff
@@ -0,0 +1,88 @@
+"""viperleed.calc.lib.rfactor.r_2."""
```
Like for r1, most comments on pendry apply also here.

michele-riva commented on src/viperleed/calc/lib/rfactor/pendry.py:
```diff
@@ -0,0 +1,115 @@
+    energy_grid : array
+        The grid on which data should be evaluated. All evaluations will ignore
+        regions in which either of the datasets is nan.
+    data_spline_1, data_spline_2 : arrays of splines
```
Are these really arrays of splines?

michele-riva commented on src/viperleed/calc/lib/rfactor/pendry.py:
```diff
@@ -0,0 +1,115 @@
+        data_2_intensity = data_spline_2(shifted_grid)
+        data_2_derivative = data_2_deriv_spline(shifted_grid)
+    else:
+        data_2_intensity, data_2_derivative, _ = data_and_derivatives_2
```
Perhaps could flatten out this nested if, like so:


```suggestion
    if data_and_derivatives_2 is None and data_spline_2 is None:
        raise TypeError('R_pendry requires either data splines or '
                        'pre-computed data_and_derivatives arrays.')
    if data_and_derivatives_2 is None:
        # when using splines, this can be sped up via CashedSplines
        data_2_deriv_spline = data_spline_2.derivative()
        # evaluate on shifted grid, allowing continuous shifts
        data_2_intensity = data_spline_2(shifted_grid)
        data_2_derivative = data_2_deriv_spline(shifted_grid)
    else:
        data_2_intensity, data_2_derivative, _ = data_and_derivatives_2
```

michele-riva replied:
Same for the first one higher up (though probably just easier to do this in the refactored function)

michele-riva commented on src/viperleed/calc/lib/rfactor/r_1.py:
```diff
@@ -0,0 +1,88 @@
+    ) / nansum_trapezoid(data_2_intensity, energy_step, axis=0)
+
+    numerators = nansum_trapezoid(
+        (data_1_intensity - beam_normalization * data_2_intensity),
```
Isn't there some `abs` needed here?

alexmiame replied:
Yes, good catch.

michele-riva commented on src/viperleed/calc/lib/rfactor/r_2.py:
```diff
@@ -0,0 +1,88 @@
+from .utils import nansum_trapezoid
+
+
+def R_2(
```
Perhaps it would be good to mention in the docstring which definition of R2 we're using. There are a couple different ones around.

michele-riva commented on src/viperleed/calc/lib/rfactor/pendry.py:
```diff
@@ -0,0 +1,115 @@
+    ):
+    """
+    Prendry R-factor, see
+    https://iopscience.iop.org/article/10.1088/0022-3719/13/5/024/meta
```
Maybe better to use doi.org? That's guaranteed to be persistent.

michele-riva commented on src/viperleed/calc/lib/rfactor/smooth.py:
```diff
@@ -0,0 +1,143 @@
+        Determines the behaviour at a minimum reaching zero intensity, see
+        https://iopscience.iop.org/article/10.1088/1361-648X/ae4af8
+    """
+    
```
no newline after function docstring

michele-riva commented on src/viperleed/calc/lib/rfactor/smooth.py:
```diff
@@ -0,0 +1,143 @@
+            data_and_derivatives_2
+            )
+    # shift data_2 to be non-negative in overlapping regions. This should fix
+    #  any potential issues cause by undershooting splines.
```
```suggestion
    #  any potential issues caused by undershooting splines.
```

michele-riva commented on src/viperleed/calc/lib/rfactor/smooth.py:
```diff
@@ -0,0 +1,143 @@
+    else:
+        data_2_intensity, data_2_derivative_1, data_2_derivative_2 = (
+            data_and_derivatives_2
+            )
```
Would this be an option (not sure if OK for JAX)?

```suggestion
        data_and_derivatives_2 = (
            data_spline_2(shifted_grid),
            data_2_deriv_1_spline(shifted_grid),
            data_2_deriv_2_spline(shifted_grid),
            )
```

(And the same for the data_1 branch)

Then the calls to `y_s` below could be simplified by unpacking via `*data_and_derivatives_2`

michele-riva replied:
same applies to rpendry

alexmiame replied:
Yes, should be fine with JAX. Packing and unpacking into/from tuples is not a problem.
The only JAX critical part here is the explicit if/else - because the JIT will only apply the path taked during the first pass.

michele-riva commented on src/viperleed/calc/lib/rfactor/smooth.py:
```diff
@@ -0,0 +1,143 @@
+    data_and_derivatives_2=None,
+    shift_2nd_spline=0.0,   # only available if passed as spline
+    alpha=4.0,
+    beta=0.15,
```
We repeat these defaults both here and in y_s. Perhaps nicer to have them as module globals?

michele-riva commented on src/viperleed/calc/lib/rfactor/smooth.py:
```diff
@@ -0,0 +1,143 @@
+    v0_imag,
+    alpha=4.0,
+    beta=0.15,
+):
```
Missing docstring

michele-riva commented on src/viperleed/calc/lib/rfactor/utils.py:
```diff
@@ -0,0 +1,113 @@
+__license__ = 'GPLv3+'
+
+import numpy as np
+from viperleed.calc.lib import dynamic_numerical_lib as dnl
```
```suggestion

from viperleed.calc.lib import dynamic_numerical_lib as dnl
```

one empty line between third-party and viperleed imports

michele-riva commented on src/viperleed/calc/lib/rfactor/utils.py:
```diff
@@ -0,0 +1,113 @@
+from viperleed.calc.lib import dynamic_numerical_lib as dnl
+
+
+def nansum_trapezoid(y, dx, axis=-1):
```
docstring missing

michele-riva commented on src/viperleed/calc/lib/rfactor/utils.py:
```diff
@@ -0,0 +1,113 @@
+    beam_correspondence : tuple
+        A tuple containing the beam correspondence, which maps the
+        experimental beams to the theoretical beams. It should be a 1D array
+        of integers with shape (n_beams,).
```
Perhaps worth mentioning that beams with the same index are averaged over. Also mention that those with -1 are removed. It may also be better to disentangle this description from the experimental beams as much as possible.

michele-riva commented on src/viperleed/calc/lib/rfactor/utils.py:
```diff
@@ -0,0 +1,113 @@
+        the beam correspondence.
+    """
+    # convert beam correspondence to numpy array
+    beam_corr = np.array(beam_correspondence, dtype=np.int32)  # force numpy
```
Perhaps worth mentioning why we force using numpy here rather than dnl.xp

michele-riva commented on src/viperleed/calc/lib/rfactor/zannazi_jona.py:
```diff
@@ -0,0 +1,133 @@
+"""viperleed.calc.lib.rfactor.zannazi_jona."""
```
See also the other modules. Most comments apply here too.

michele-riva commented on src/viperleed/calc/lib/rfactor/zannazi_jona.py:
```diff
@@ -0,0 +1,133 @@
+    ):
+    """
+    Zannazi-Jona R-factor, see
+    https://www.sciencedirect.com/science/article/pii/0039602877904289
```
use doi

michele-riva commented on src/viperleed/calc/lib/rfactor/zannazi_jona.py:
```diff
@@ -0,0 +1,133 @@
+        1 / nansum_trapezoid(data_1_intensity, energy_step, axis=0) / 0.027
+    )
+
+    # # calculate normalization for each beam
```
```suggestion
    # calculate normalization for each beam
```

michele-riva commented on src/viperleed/calc/lib/rfactor/zannazi_jona.py:
```diff
@@ -0,0 +1,133 @@
+        data_2_intensity, data_2_derivative_1, data_2_derivative_2 = (
+            data_and_derivatives_2
+            )
+    mask_1 = dnl.xp.isnan(data_1_intensity)
```
Is there a reason to have two separate masks? Looks like taking the logical_or from the beginning would be the same. Also, it may be nice to give it a more evocative name, as I got confused toward the end of the function as to what was `True` and what was `False` within `mask`. (E.g., `mask` -> `no_overlap`)

michele-riva commented on src/viperleed/calc/lib/dynamic_numerical_lib.py:
```diff
@@ -0,0 +1,152 @@
+"""viperleed.calc.lib.dynamic_numerical_lib.
```
```suggestion
"""Module dynamic_numerical_lib of viperleed.calc.lib.
```

michele-riva commented on src/viperleed/calc/lib/dynamic_numerical_lib.py:
```diff
@@ -0,0 +1,152 @@
+To swap the numerical library in use, set the corresponding variable
+from this module.
+
+For example, to use JAX instead of NumPy, you can do:
+
+    from viperleed.calc.lib import dynamic_numerical_lib as dnl
+    import jax
+    dnl.xp = jax.numpy
```
This comment seems to imply that this is enough to switch. However, splines should also be replaced for full JAX support, as should the various numpy implementations below.

michele-riva commented on src/viperleed/calc/lib/dynamic_numerical_lib.py:
```diff
@@ -0,0 +1,152 @@
+__license__ = 'GPLv3+'
+
+import numpy as _np  # NumPy is the default numerical library
+from scipy.interpolate import (
+    CubicSpline as _SciCubicSpline,
+)
+from scipy.interpolate import (
+    PPoly as _SciPPoly,
+)
+
+xp = _np
+CubicSpline = _SciCubicSpline
+PPoly = _SciPPoly
```
It is probably enough to just do


```suggestion
from scipy.interpolate import CubicSpline
from scipy.interpolate import PPoly

xp = _np
```

Unless we need access to the scipy splines in this module (which, at present, does not seem to be the case)

alexmiame replied:
It is enough, but this is intentionally explicit to draw attention to it.

michele-riva commented on src/viperleed/calc/lib/dynamic_numerical_lib.py:
```diff
@@ -0,0 +1,152 @@
+
+    This is a wrapper around `dnl.xp.bincount` to provide a consistent
+    interface across different numerical libraries. Some libraries may
+    use `minlength` instead of `length` as a parameter name.
```
This feels inconsistent with the implementation that always uses numpy.

alexmiame replied:
It is inconsitent; thus the need to patch it.

michele-riva replied:
I meant that the docstring is inconsistent with the implementation: it says it is a wrapper around dnl.xp.bincount, but it uses numpy.bincount irrespective of what dnl.xp is, so it's not a wrapper around dnl.xp.bincount but a wrapper around numpy.bincount.

michele-riva commented on src/viperleed/calc/lib/dynamic_numerical_lib.py:
```diff
@@ -0,0 +1,152 @@
+    ----------
+    func : callable
+        Function to map.
+    in_axes : int | None | tuple
```
```suggestion
    in_axes : int or None or tuple
```

michele-riva commented on src/viperleed/calc/lib/dynamic_numerical_lib.py:
```diff
@@ -0,0 +1,152 @@
+    return _np.bincount(x, weights=weights, minlength=length)
+
+
+def vmap(func, in_axes=0, out_axis=0):
```
`jax.vmap` has `out_axes` instead of `out_axis` as argument name. Also, reading jax.vmap docstring it sounds as if out_axes is more complex than just stacking things. Do we even need this argument?

alexmiame replied:
It is more complex of course; this is only intended as a simple patch that is "good enough" for internal purposes.

michele-riva replied:
Then perhaps it would be a good idea to clarify this in the docstring

michele-riva commented on src/viperleed/calc/lib/dynamic_numerical_lib.py:
```diff
@@ -0,0 +1,152 @@
+        Function to map.
+    in_axes : int | None | tuple
+        For each positional arg: which axis to map over, or None to broadcast.
+        If an int is given, it applies to the first arg and all others are None.
```
I may interpreting wrongly the documentation for jax.vmap, but this seems not the way JAX does it.

michele-riva commented on src/viperleed/calc/lib/spline_interpolation.py:
```diff
@@ -0,0 +1,86 @@
+__created__ = '2024-02-19'
+__license__ = 'GPLv3+'
+
+
+import numpy as np
+from viperleed.calc.lib import dynamic_numerical_lib as dnl
```
```suggestion
import numpy as np

from viperleed.calc.lib import dynamic_numerical_lib as dnl
```

michele-riva commented on src/viperleed/calc/lib/spline_interpolation.py:
```diff
@@ -0,0 +1,86 @@
+        self._last_y = None
+
+    def __call__(self, x):
+        if x is self._last_x:
```
Is `is` the right comparison here? Seems restrictive compared to `==` which should give the same result (unless I misunderstand)

michele-riva commented on src/viperleed/calc/lib/spline_interpolation.py:
```diff
@@ -0,0 +1,86 @@
+    __slots__ = ("_s", "_deriv_cache", "_last_x", "_last_y")
+
+    def __init__(self, spline):
+        self._s = spline
```
Perhaps `_spline` would be more obvious?

michele-riva commented on src/viperleed/calc/lib/spline_interpolation.py:
```diff
@@ -0,0 +1,86 @@
+        return self._deriv_cache[key]
+
+    def __getattr__(self, name):
+        # delegate anything else (e.g., extrapolate, c, x, etc.)
```
This comment is not so clear unless readers know that `__getattr__` is only invoked if self does not have the attribute explicitly defined.

michele-riva commented on src/viperleed/calc/lib/spline_interpolation.py:
```diff
@@ -0,0 +1,86 @@
+):
+    """Construct a piecewise cubic spline interpolator with ragged edges.
+
+    The interpolator uses a cubic spline to interpolate data.
```
Missing Parameters and Returns docstring

michele-riva commented on src/viperleed/calc/lib/spline_interpolation.py:
```diff
@@ -0,0 +1,86 @@
+    return subarray_spline, start_index
+
+
+def interpolate_ragged_array(
+    x, y, axis=0, bc_type='not-a-knot', extrapolate=False
+):
```
Missing docstring

michele-riva commented on src/viperleed/calc/lib/spline_interpolation.py:
```diff
@@ -0,0 +1,86 @@
+        if all_nans:
+            continue
+        spline, start_id = make_1d_ragged_cubic_spline(
+            x, y[:, dim], axis=0, bc_type=bc_type, extrapolate=None
```
```suggestion
            x, _y, axis=0, bc_type=bc_type, extrapolate=None
```

michele-riva commented on src/viperleed/calc/lib/spline_interpolation.py:
```diff
@@ -0,0 +1,86 @@
+
+
+def interpolate_ragged_array(
+    x, y, axis=0, bc_type='not-a-knot', extrapolate=False
```
`axis` and `extrapolate` seem unused

michele-riva commented on src/viperleed/calc/sections/rfactor.py:
```diff
@@ -17,23 +17,15 @@
-from viperleed.calc.lib import fs_utils
-from viperleed.calc.lib import leedbase
+from viperleed.calc.files.iorfactor import (beamlist_to_array,
+                                            write_Rfactorpdf_from_splines)
```
perhaps not useful to import specific names from iorfactor that we already import higher up?

michele-riva commented on src/viperleed/calc/sections/rfactor.py:
```diff
@@ -126,299 +118,148 @@ def _fetch_and_check_spectra(rp, index, name):
+            'interpolation (INTPOL_DEG=3).'
         )
+        rp.setHaltingLevel(1)
+        return []
```
Perhaps `raise NotImplementedError` would be more explicit? Alternatively, I would force a fallback to cubic rather than not calculating any R factor at all.

Probably this info should also go to the doc of INTPOL_DEG.

michele-riva commented on src/viperleed/calc/sections/rfactor.py:
```diff
@@ -126,299 +118,148 @@ def _fetch_and_check_spectra(rp, index, name):
+                shift_2nd_spline=shift,
+                groups=None,
+            )
+        )
```
Perhaps a list comprehension would be better than appending

michele-riva commented on src/viperleed/calc/sections/rfactor.py:
```diff
@@ -600,3 +441,13 @@ def run_legacy_rfactor(sl, rp, for_error, name, theobeams, index, only_vary):
+def determine_integer_or_fractional(rp):
+    """Determine whether beams are integer or fractional."""
+    iorf = []
+    for i, beam in enumerate(rp.expbeams):
```
`i` (and thus `enumerate`) unused. Could be done with a comprehension.

michele-riva commented on src/viperleed/calc/sections/rfactor.py:
```diff
@@ -600,3 +441,13 @@ def run_legacy_rfactor(sl, rp, for_error, name, theobeams, index, only_vary):
+    """Determine whether beams are integer or fractional."""
+    iorf = []
+    for i, beam in enumerate(rp.expbeams):
+        if beam.hk[0] % 1.0 != 0.0 or beam.hk[1] % 1.0 != 0.0:
```
```suggestion
        if beam.hk[0] % 1.0 or beam.hk[1] % 1.0:
```

michele-riva commented on src/viperleed/calc/from_ase.py:
```diff
@@ -465,17 +461,57 @@ def _write_poscar(slab, exec_path):
+def _match_beams_by_label(expbeams, theobeams):
+    """
+    Return matched exp/theory beam lists using direct hkfrac matching.
+    """
```
```suggestion

def _match_beams_by_label(expbeams, theobeams):
    """Return matched exp/theory beam lists using direct hkfrac matching."""
```

michele-riva commented on src/viperleed/calc/from_ase.py:
```diff
@@ -465,17 +461,57 @@ def _write_poscar(slab, exec_path):
 ):
-    """Compute the Pendry R-factor between two CSV files (in ViPErLEED format).
+    """
+    Compute the Pendry R-factor between two CSV files (in ViPErLEED format).
```
```suggestion
    """Compute the R-factor between two CSV files (in ViPErLEED format).
```

michele-riva commented on src/viperleed/calc/from_ase.py:
```diff
@@ -488,80 +524,81 @@ def rfactor_from_csv(
         the CSV contents rather than a path to the file. For reading,
         a StringIO will be used internally.
-    v0r_shift_range : Sequence of two float, default=(-3, 3)
+    v0r_shift_range : tuple of two float, default=(-3, 3)
```
Any reason not to accept a `Sequence`?

michele-riva commented on src/viperleed/calc/from_ase.py:
```diff
@@ -488,80 +524,81 @@ def rfactor_from_csv(
-    return_beam_arrays : bool, default = False
-        For debugging, you can optionally return the interpolated beams,
-        y-functions, number of overlapping points & per-beam R-factors.
+        the minimum and maximum shifts (given in eV).
```
The note about being integer multiples of `intpol_step` still applies. I think this is enforced by the `IVShiftRange` class.

michele-riva commented on src/viperleed/calc/from_ase.py:
```diff
@@ -488,80 +524,81 @@ def rfactor_from_csv(
-    which_r = 1
+    if intpol_deg is not None:
+        _LOGGER.warning(
+            "intpol_deg is deprecated in rfactor_from_ase(); "
```
```suggestion
            "intpol_deg is deprecated in rfactor_from_csv(); "
```

michele-riva replied:
(more occurrences below)

michele-riva commented on src/viperleed/calc/from_ase.py:
`ValueError` seems more appropriate here. Also, is this check really useful?

michele-riva commented on src/viperleed/calc/from_ase.py:
```diff
@@ -31,18 +30,15 @@
 from viperleed.calc.files import parameters, poscar
 from viperleed.calc.files.beams import readOUTBEAMS
 from viperleed.calc.files.ivplot import plot_iv
+from viperleed.calc.files.iorfactor import beamlist_to_array
+from viperleed.calc.files.iorfactor import write_Rfactorpdf_from_splines
```
module imported as rf_io higher up. Can take attributes from it rather than importing more stuff

michele-riva commented on src/viperleed/calc/from_ase.py:
```diff
@@ -571,181 +608,118 @@ def rfactor_from_csv(
+    rpars.V0_IMAG = v0i
+
+    theo_grid, _, _, theo_intensities = beamlist_to_array(theobeams)
+    rpars.THEO_ENERGIES = TheoEnergies.from_sorted_grid(theo_grid)
```
A large fraction of what is currently done here also happens identically in sections.rfactor. Perhaps it would be a good idea to factor out stuff correspondingly?

fkraushofer committed:
chore: fix error message

fkraushofer committed:
fix: avoid divide-by-zero in y_s

y_s currently prints divide-by-zero warnings if the second derivative is zero anywhere. There was no true bug, as these values were later filtered out ('condition'), but to avoid spamming the log, it's nicer to filter this already earlier on.

fkraushofer committed:
feat: R analysis update, use new y-functions

- re-implement Rfactor_analysis.pdf for spline-based R-factor
- refactor old Rfactor pdf function to also use the spline implementation for y-functions
- deprecate leedbase.getYfunc, use the new lib.rfactor y-function implementations instead
- add which_r parameter to distinguish R_P and R_S

michele-riva commented on src/viperleed/calc/files/iorfactor.py:
```diff
@@ -23,12 +23,15 @@
 from viperleed.calc.files.beams import writeAUXEXPBEAMS
 from viperleed.calc.files.ivplot import plot_iv
-from viperleed.calc.lib import leedbase
+from viperleed.calc.lib import leedbase, spline_interpolation
```
```suggestion
from viperleed.calc.lib import leedbase
from viperleed.calc.lib import spline_interpolation
```

I tried to keep one import per line to prevent missing imports when reading.

michele-riva commented on src/viperleed/calc/files/iorfactor.py:
```diff
@@ -705,6 +733,9 @@ def writeRfactorPdf(beams, colsDir='', outName='Rfactor_plots.pdf',
             layout will be adapted automatically. Numbers that are not nicely
             divisible may be rounded up, resulting in some whitespace.
             default: 2 (one column, two rows).
+    which_r : str
```
Can also be an RFactorType, I guess

michele-riva commented on src/viperleed/calc/files/iorfactor.py:
```diff
@@ -719,42 +750,28 @@ def writeRfactorPdf(beams, colsDir='', outName='Rfactor_plots.pdf',
-    except PermissionError:
-        logger.error("writeRfactorPdf: Cannot open file {}. Aborting."
-                     .format(analysisFile))
+    if which_r != 'pendry' and which_r != 'smooth':
```
```suggestion
    if which_r not in {'pendry', 'smooth'}:
```

michele-riva commented on src/viperleed/calc/files/iorfactor.py:
```diff
@@ -719,42 +750,28 @@ def writeRfactorPdf(beams, colsDir='', outName='Rfactor_plots.pdf',
+    _write_rfactor_analysis_pdf_from_splines(
+        theo_spline, exp_spline,  out_grid,
+        xyTheo, xyExp,
+        labels, rfacs,
```
I'd suggest only one arg per line in this call, otherwise it's easy to miss some.

michele-riva commented on src/viperleed/calc/files/iorfactor.py:
```diff
@@ -966,6 +917,209 @@ def beamlist_to_array(beams):
+        _write_rfactor_analysis_pdf_from_splines(
+            theo_spline, exp_spline, out_grid,
+            xy_theo, xy_exp,
+            labels, r_fac_per_beam,
```
like above, one arg per line

michele-riva commented on src/viperleed/calc/files/iorfactor.py:
```diff
@@ -966,6 +917,209 @@ def beamlist_to_array(beams):
+            v0i=rpars.V0_IMAG,
+            formatting=rpars.PLOT_IV,
+            which_r=rpars.R_FACTOR_TYPE,
+        )
+
+def _yfuncs_from_splines(exp_spline, theo_spline, out_grid, v0i, which_r):
```
```suggestion


def _yfuncs_from_splines(exp_spline, theo_spline, out_grid, v0i, which_r):
```

michele-riva commented on src/viperleed/calc/files/iorfactor.py:
```diff
@@ -966,6 +917,209 @@ def beamlist_to_array(beams):
+    exp_deriv_1_spline = exp_spline.derivative()
+    exp_deriv_2_spline = exp_deriv_1_spline.derivative()
+    exp_derivative_1 = np.asarray(exp_deriv_1_spline(out_grid))
+    exp_derivative_2 = np.asarray(exp_deriv_2_spline(out_grid))
```
Perhaps we can reuse some of my suggested refactoring in rfactor here?

Also, are the conversions to numpy arrays needed? I guess the other option is for these to be jax.numpy arrays

alexmiame replied:
Conversion should not be needed

michele-riva commented on src/viperleed/calc/files/iorfactor.py:
```diff
@@ -966,6 +917,209 @@ def beamlist_to_array(beams):
+        )
+    raise ValueError(
+        f'Cannot write R-factor analysis PDF for R_FACTOR_TYPE={which_r!r}.'
+    )
+
+def _rfactor_columns_to_spline(xy):
```
```suggestion


def _rfactor_columns_to_spline(xy):
```

michele-riva commented on src/viperleed/calc/files/iorfactor.py:
```diff
@@ -966,6 +917,209 @@ def beamlist_to_array(beams):
+    for i, beam_xy in enumerate(xy):
+        if len(beam_xy) == 0:
+            continue
+        beam_xy = np.asarray(beam_xy, dtype=float)
```
Is this necessary?

michele-riva commented on src/viperleed/calc/files/iorfactor.py:
```diff
@@ -966,6 +917,209 @@ def beamlist_to_array(beams):
+        y=intensities,
+        bc_type='not-a-knot',
+    )
+    return spline_interpolation.CachedSpline(spline), grid
+
+@skip_without_matplotlib
+def _write_rfactor_analysis_pdf_from_splines(
```
```suggestion


@skip_without_matplotlib
def _write_rfactor_analysis_pdf_from_splines(
```

michele-riva commented on src/viperleed/calc/lib/rfactor/smooth.py:
```diff
@@ -0,0 +1,145 @@
+    beta=0.15,
+):
+    positive_d2 = second_derivative > 0
+    safe_d2 = dnl.xp.where(positive_d2, second_derivative, 1.0)
```
Perhaps worth commenting here that the value 1.0 does not really matter as safe_d2 enters the final Y value (through conditional_denominator) only where the second derivative is positive.

michele-riva commented on src/viperleed/calc/sections/rfactor.py:
```diff
@@ -17,23 +17,15 @@
-from viperleed.calc.constants import DEFAULT_TENSORS
-from viperleed.calc.files import iorfactor
-from viperleed.calc.files import iotensors
+from viperleed.calc.constants import DEFAULT_OUT, DEFAULT_SUPP, DEFAULT_TENSORS
+from viperleed.calc.files import iorfactor, iotensors
```
Not a fan of the flattening of these imports that comes from Ruff. The point of having one import per line was to (i) limit the number of lines that are affected by a commit should one add more imports, and (ii) not risk missing imports when reading quickly over them. (Also applies further below.)

alexmiame replied:
As dicussed several times before, this is currently not a configurable setting in ruff and using consistent formatting outweighs this **by far, imo**.
Anyone is free to implement a different auto-formatter and linter for the project, but I believe our resources are better spent eleswhere.

michele-riva commented on src/viperleed/calc/sections/rfactor.py:
```diff
@@ -126,299 +118,148 @@ def _fetch_and_check_spectra(rp, index, name):
 
 
 def run_new_rfactor(sl, rp, for_error, name, theobeams, expbeams):
```
Missing docstring

michele-riva commented on src/viperleed/calc/sections/rfactor.py:
```diff
@@ -596,7 +437,18 @@ def run_legacy_rfactor(sl, rp, for_error, name, theobeams, index, only_vary):
             logger.warning("Error plotting R-factors.", exc_info=True)
     return rfaclist
+
+def determine_integer_or_fractional(rp):
```
```suggestion


def determine_integer_or_fractional(rp):
```

michele-riva replied:
Perhaps also good to do rp -> rpars

michele-riva commented on src/viperleed/calc/from_ase.py:
```diff
@@ -571,181 +608,118 @@ def rfactor_from_csv(
-    return best_r, best_v0r
+    if plot_iv is not None:
+        rpars.PLOT_IV.update(plot_iv)
+        rpars.expbeams = expbeams
```
Why re-assigning these? Looks like it is done already higher up

michele-riva commented on src/viperleed/calc/from_ase.py:
```diff
@@ -571,181 +608,118 @@ def rfactor_from_csv(
 
+    return best_r, best_v0r, r_beams
 
 def plot_iv_from_csv(
```
Is this function still useful and up to date?

michele-riva commented on tests/calc/lib/rfactor/test_rfactors_unit.py:
```diff
@@ -0,0 +1,95 @@
+        data_spline_1=sin_spline,
+        data_spline_2=cos_spline,
+    )
+    assert calc_r == pytest.approx(1.9047921884401726, rel=1e-6)
```
```suggestion
    assert calc_r == pytest.approx(1.904792, rel=1e-6)
```

michele-riva commented on tests/calc/lib/rfactor/test_rfactors_unit.py:
```diff
@@ -0,0 +1,95 @@
+    assert result.ndim == 0 or result.shape == ()  # scalar spline case
+
+
+def test_pendry_matches_existing_reference_value(sin_spline, cos_spline):
```
May want to use the synthetic_pair fixture here too? There would be no need to manually re-define v0i, energies, and the step

michele-riva commented on tests/calc/lib/rfactor/conftest.py:
```diff
@@ -0,0 +1,197 @@
+"""Shared fixtures and helpers for rfactor tests."""
+
+from __future__ import annotations
```
Annotations not needed.

michele-riva commented on tests/calc/lib/rfactor/conftest.py:
```diff
@@ -0,0 +1,197 @@
+"""Shared fixtures and helpers for rfactor tests."""
+
+from __future__ import annotations
+
```
Missing: module dunders with authors, copyright, etc.

Applies to all of these test modules

michele-riva reviewed (changes requested):
Overall looks good. I have the usual bunch of suggestions, though.

michele-riva reviewed (commented):
Overall looks good. I have the usual bunch of suggestions, though.

Also, this seems to have some merge conflicts that need resolution

fkraushofer committed:
minor,chore: batch submit minor changes from review

Co-authored-by: Michele Riva <riva@iap.tuwien.ac.at>

fkraushofer committed:
Merge branch '0.15.0' into python_rfactor

fkraushofer commented:
Merge conflicts resolved in b2e9e3c1081f1548c6e5e20f56d2fcd5ae7e6faa (were due to the `R_FACTOR_SMOOTHING` rename)

After a quick look at the requested changes, it seems to be a mix between stuff from @alexmiame and me - we should coordinate.

fkraushofer committed:
minor,chore: more minor changes

Co-authored-by: Michele Riva <riva@iap.tuwien.ac.at>

alexmiame committed:
fix, doc: docstring

alexmiame committed:
chore: use properties

alexmiame committed:
chore: ValueError for unknown R factor variation

alexmiame committed:
doc,minor

alexmiame committed:
doc: clarify

alexmiame committed:
style,minor

alexmiame committed:
style: var name and line breaks

alexmiame committed:
chore: remove unneccessary conversion to numpy arrays

alexmiame committed:
doc,minor: use doi.org

alexmiame committed:
doc,minor

alexmiame committed:
feat: more numerically stable version

as suggested by @michele-riva

alexmiame committed:
doc: add docstrings for pendry R functions

alexmiame committed:
doc: add raises doc for R factors

alexmiame committed:
doc: docstring first lines

alexmiame committed:
doc: use doi.org link instead of iopscience

alexmiame committed:
style: move default alpha and beta to module global

alexmiame committed:
doc: add note

Suggested by @michele-riva

alexmiame committed:
doc: add y_s docstring

alexmiame committed:
doc: missing raises doc

alexmiame committed:
feat: lower case R

as discussed in #500

alexmiame committed:
feat: lower case r_zj

alexmiame committed:
minor: lowercase in error msg

alexmiame committed:
style,minor: whitespace

alexmiame committed:
doc: add/clarify dnl docstrings

alexmiame committed:
fix: argument name

alexmiame committed:
doc: add into to docstrings and comments

alexmiame committed:
doc: use doi.org for ZJ reference

alexmiame committed:
feat,refactor: NaN masking for ZJ R factor

alexmiame committed:
doc: add ref to which R2

alexmiame committed:
doc: clarify docstring

alexmiame committed:
doc: clarify grouped R factor NaN behaviour

alexmiame committed:
style,minor

alexmiame committed:
doc: change docstring

alexmiame committed:
doc: add kwargs to doc

alexmiame committed:
doc: more specific

alexmiame committed:
feat: utility to coerce data to energy grid

alexmiame committed:
fix: truncate data that is passed through

alexmiame committed:
style: simpler eval

alexmiame committed:
feat: use coerce to grid util in pendry

alexmiame committed:
fix,test: update expected R factor values

The updated masking behaviour in ac1c3feec5dda6117d96d738319c1b055a35d22e
slightly alters the R factor calculation (order of 10e-4 in R factor).
This requires the expected R factors to be updated.
@michele-riva @fkraushofer could you have a quick look to make sure I'm not missing anything?

alexmiame committed:
style,minor: alphabetic sorting

alexmiame committed:
chore: remove unused import

alexmiame committed:
fix: undo accidental removal of shift non-negative

alexmiame committed:
feat: use coerce to grid for smooth R

alexmiame committed:
Revert "fix,test: update expected R factor values"

This reverts commit c98f6a2d36629962e8812bf57bdeb587de43a3eb.

alexmiame committed:
feat: use coerce to grid for R2

alexmiame committed:
feat: use coerce to grid for R1

alexmiame committed:
feat: use coerce to grid for RZJ

alexmiame committed:
fix: missing abs for R1

alexmiame committed:
fix,test: update expected R1 factor values

alexmiame committed:
style,minor: import order

alexmiame committed:
style: rename attrib

alexmiame committed:
style: use defined var

alexmiame committed:
doc: add parameters and return

alexmiame committed:
feat: use == comparison

alexmiame committed:
fix: pass args

alexmiame committed:
doc: docstring for function

alexmiame committed:
doc: clarify comment

alexmiame committed:
style: unused var

alexmiame committed:
style: make loop into list comprehension

alexmiame committed:
style: avoid double import
